main.go

```diff
@@ -181,7 +183,10 @@ func setupReconcilers(ctx context.Context, mgr ctrl.Manager) {
 		RancherClient:      rancherClient,
 		WatchFilterValue:   watchFilterValue,
 		InsecureSkipVerify: insecureSkipVerify,
-	}).SetupWithManager(ctx, mgr, controller.Options{MaxConcurrentReconciles: concurrencyNumber}); err != nil {
+	}).SetupWithManager(ctx, mgr, controller.Options{
+		MaxConcurrentReconciles: concurrencyNumber,
+		CacheSyncTimeout:        maxDuration,
```

Comment:
Cache setup under the hood is already a polling process, with a very detailed logging to it. I don’t understand why there should be an additional polling logic on top where we could utilize existing functionality upstream. Here are the logs the user would see from a rancher-turtles container, when the rancher CRDs are not installed:
```

│ E1004 14:42:20.654299       1 kind.go:68] controller-runtime/source/EventHandler "msg"="failed to get informer from cache" "error"="failed to get API │
│  group resources: unable to retrieve the complete list of server APIs: provisioning.cattle.io/v1: the server could not find the requested resource"   │
│ E1004 14:42:30.654520       1 kind.go:68] controller-runtime/source/EventHandler "msg"="failed to get informer from cache" "error"="failed to get API │
│  group resources: unable to retrieve the complete list of server APIs: provisioning.cattle.io/v1: the server could not find the requested resource"   │
│ E1004 14:42:40.654445       1 kind.go:68] controller-runtime/source/EventHandler "msg"="failed to get informer from cache" "error"="failed to get API │
│  group resources: unable to retrieve the complete list of server APIs: provisioning.cattle.io/v1: the server could not find the requested resource"
```
As you can see, there is a log message with a 10s periodicity which says that this specific rancher CRD is not available. Internally the cache implementation is evaluated lazily, so there is no danger that this constant polling will overload api server, as it is not doing a full API version discovery, but only asks for a single CRD definition at a time.

Reply:
As for disabling the cache timeout in general - this is only a temporal measure, which will no longer be applied to the CAPI resources in the future. Once the Rancher resources will be managed by a separate cluster definition, this specific setting could be delegated to the second cluster setup.

Reply:
To me and you the log messages mean something but they aren't overly helpful generally. I was thinking of something along these lines:

```
[Preflight] Running checks
[Preflight] Checking for Rancher CRDs
[Preflight] ERROR: Failed check for Rancher CRDs. Please install Rancher to use this extension
....
```

Reply:
And then in the future we could extend it to also check the "embedded capi" feature flag....if we wanted to.

Reply:
These are just suggestions, if consensus is that using the cache sync timeout is sufficient then i'm good with that.

Reply:
Should this PR and attached issue be closed considering that https://github.com/rancher-sandbox/rancher-turtles/issues/185#issuecomment-1747264384 is adding expectation for Rancher cluster be provisioned first in any installation scenario, + it is the process in the docs?

Reply:
No i don't think so. The docs cover the early access which will not cover all scenarios we may need to support in the future. This gives us options for the future.

Reply:
I don't like error messages in logs but at the same time, if we can avoid writing something custom, I'd like to go that way. Less code, less maintenance :) I also think such preflight checks should be implemented on the controller-runtime side

Reply:
For the preflight checks to be implement on the controller-runtime side it would need a "hook" mechanism for everything we'd ideally want to check. For example, we may want to check that the feature flag is disabled before starting the controllers. Interesting idea upstreaming the pre-flight functionality though.

I don't think using the cache sync timeout is the right solution to the problem, it gives a poor UX in my opinion. However, i don't feel strongly enough on this....so @Danil-Grigorev if there is consensus on this change (i.e. via the approvals) then go with this 👍

Reply:
Meanwhile I also agree this might not be the best solution to the issue, I think we could implement a custom controller without upstreaming it in controller runtime, something like CAPI operator has for [preflight checks](https://github.com/kubernetes-sigs/cluster-api-operator/blob/2ef0046c4aea606cba2cd39319484d5b10403b5d/internal/controller/preflight_checks.go#L53)

Reply:
This preflight check should be called validation check, and we are not doing a validation for a resource here @furkatgofurov7